README.md
# Disqus plugin for Craft CMS

Integrate the [Disqus](https://disqus.com) commenting system into [Craft CMS](http://buildwithcraft.com) websites, including Single Sign On (SSO) and custom login/logout URLs.

Related: [Disqus for Craft 3.x](https://github.com/nystudio107/craft3-disqus)

**Installation**

1. Download & unzip the file and place the `disqus` directory into your `craft/plugins` directory
2.  -OR- do a `git clone https://github.com/nystudio107/disqus.git` directly into your `craft/plugins` folder.  You can then update it with `git pull`
3.  -OR- install with Composer via `composer require nystudio107/disqus`
4. Install plugin in the Craft Control Panel under Settings > Plugins
5. The plugin folder should be named `disqus` for Craft to see it.  GitHub recently started appending `-master` (the branch name) to the name of the folder for zip file downloads.

## Configuring Disqus

First, make sure you have [set up a Disqus account](https://disqus.com/websites/).

Next in the Craft Admin CP, go to Settings->Plugins->Disqus and enter the Short Name for your Disqus site.  This is the only required setting for the Disqus plugin.

All settings are also configurable via the `config.php` file, which is a multi-environment friendly way to store the default settings.  Don't edit this file, instead copy it to `craft/config` as `disqus.php` and make your changes there.

### Single Sign On (SSO)

The real usefulness of the Disqus plugin is that it takes care of the Single Sign On (SSO) integration with your Craft site.

Before you can use this, you'll need to set up the Disqus SSO API as described on the [Integrating Single Sign-On](https://help.disqus.com/customer/portal/articles/236206-integrating-single-sign-on) web page.

Then copy and paste the API Key and API Secret into the Disqus plugin settings, and turn on the "User Single Sign On" lightswitch.

### Custom Login/Logout URLs

The Disqus plugin will also take care of the custom login/logout URLs, should you wish to use them.  Please see [Adding your own SSO login and logout links](https://help.disqus.com/customer/portal/articles/236206-integrating-single-sign-on#sso-login) for details.

You only need this is you want to have a custom login button displayed in the Disqus UI itself.  

`url` should be the address of your login page. The page will be opened in a new window and it must close itself after authentication is done. That is how we know when it is done and reload the page.

`logout` should be set to `http://example.com/actions/disqus/logoutRedirect` to hit the Disqus controller that handles the logout and redirect.

## Using the Disqus plugin in your templates

All of these methods accomplish the same thing:

    {# Output the Disqus embed code using the 'disqusEmbed' function #}
    {{ disqusEmbed( DISQUS_IDENTIFIER, DISQUS_TITLE, DISQUS_URL, DISQUS_CATEGORY_ID, DISQUS_LANGUAGE) }}

    {# Output the Disqus embed code using the 'disqusEmbed' filter #}
    {{ DISQUS_IDENTIFIER | disqusEmbed(DISQUS_TITLE, DISQUS_URL, DISQUS_CATEGORY_ID, DISQUS_LANGUAGE) }}

    {# Output the Disqus embed code using the 'disqusEmbed' variable #}
    {{ craft.disqus.disqusEmbed( DISQUS_IDENTIFIER, DISQUS_TITLE, DISQUS_URL, DISQUS_CATEGORY_ID, DISQUS_LANGUAGE) }}

All of the parameters except for `DISQUS_IDENTIFIER` are optional.  For more information on what these parameters are, please see [Javascript configuration variables](https://help.disqus.com/customer/portal/articles/472098-javascript-configuration-variables)

Disqus ignores any settings that are empty strings, e.g.: `''`

The typical Twig tag you'll use would look like this:

    {{ disqusEmbed(entry.slug, entry.title, entry.url) }}

...which will result in comments that are unique on a per-entry basis.

In its most basic case, this will result in output to your Craft template that looks like this:

    <div id="disqus_thread"></div>
    <script data-cfasync="false" type="text/javascript">
        /**
         *  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
         *  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables
         */

        if (typeof disqus_config !== 'undefined') {
            var _old_disqus_config = disqus_config;
        }
        var disqus_config = function() {
            if (typeof _old_disqus_config !== 'undefined') {
                _old_disqus_config.apply(this);
            }
            this.page.url = 'DISQUS_URL';
            this.page.identifier = 'DISQUS_IDENTIFIER';
            this.page.title = 'DISQUS_TITLE';
            this.page.category_id = 'DISQUS_CATEGORY_ID';
            this.language = 'DISQUS_LANGUAGE';
        };

        (function() {  // REQUIRED CONFIGURATION VARIABLE: EDIT THE SHORTNAME BELOW
            var d = document, s = d.createElement('script');

            s.src = '//DISQUS_SHORTNAME.disqus.com/embed.js';  // IMPORTANT: Replace EXAMPLE with your forum shortname!

            s.setAttribute('data-timestamp', +new Date());
            (d.head || d.body).appendChild(s);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>

The `DISQUS_SHORTNAME` setting is taken from the Admin CP or `config.php` settings, and the rest of the `DISQUS_*` settings are passed in as variables from the `disqusEmbed` Twig filter/function.

If you have turned on "User Single Sign On" it will also output something like this inside the above tag:

    this.page.remote_auth_s3 = "eyJpZCI6IjEiLCJ1c2VybmFtZSI6IkFkbWluIiwiZW1haWwiOiJhbmRyZXdAbWVnYWxvbWFuaWFjLmNvbSJ9 c0e4b8f2eca3c0e995cdd64ba2dedd720820ab5b 1431214361";
    this.page.api_key = "GTX1r1JBbiJah3hzZkBO06hI71VxjyWxgdurckHYBWLiELkHDidVmnDkBW0XeROe";

Which, assuming you've set up the Disqus SSO properly, will allow your Craft users to be logged into Disqus using your Craft website credentials.

If you have "Use Custom Login/Logout URLs" turned on, it will also generate the `this.sso` settings for you, [as described here](https://help.disqus.com/customer/portal/articles/236206-integrating-single-sign-on#sso-login)

## Disqus Multi-lingual websites

By default, Disqus will use the language you have set in `Disqus Admin > Setup > Appearance`, however you can use it on [Multi-lingual websites](https://help.disqus.com/customer/portal/articles/466249-multi-lingual-websites) as well.

The `DISQUS_LANGUAGE` parameter you can provide to `{{ disqusEmbed() }}` allows you to control the language that the Disqus embed is displayed in. The comments, however, will still be the same for all languages.

If you wish to have the comments themselves be different per-language, you can do something like:

    {{ disqusEmbed(entry.slug ~ "_" ~ entry.locale, entry.title, entry.url, '', entry.locale ) }}

This will result in comments that are different for each language, and the Disqus embed will be displayed in the same language as the comments.

Brought to you by [nystudio107](http://nystudio107.com)


CHANGELOG.md
# Disqus Changelog

## 1.1.1 - 2017.03.20
### Changed
* Fixed broken SSO implementation by moving it all to one template
* Changed `[]` to `array()` for ancient PHP compatibility
* Deprecated the separate `outputSSOTag()` functions
* Removed errant (and unneeded) `|escape('js')` filters
* Updated `README.md`

## 1.1.0 - 2017.03.18
### Added
* All settings can now be overridden via `config.php` in a multi-environment friendly way
* Added multi-lingual support for both the Disqus embed and comments
* Added Composer support
* Broke out the changelog into `CHANGELOG.md`

### Changed
* Separated the various tags out into Twig templates
* Cleaned up & refactored the templates & code

### 1.0.3 - 2016.10.03
 
 * [Fixed] Resolve broken $disqusIdentifier encoding when set as a string.
 * [Fixed] Fixes Disqus from loading when encoding the URL
 * [Improved] Updated README.md
 
 ### 1.0.2 - 2016.09.19
 
 * [Improved] We now JSON-encode the data in the Disqus embed
 * [Improved] Added data-cfasync=false to the script tags for CloudFlare RocketScript support
 * [Improved] Updated README.md
 
 ### 1.0.1 - 2015.11.23
 
 * Added support for Craft 2.5 new plugin features
 * Added a controller to handle the custom logout URL
 * Fixed an issue where custom avatars no longer appeared
 
 ### 1.0.0 - 2015.05.09
 
 * Initial release

Brought to you by [nystudio107](http://nystudio107.com)

DisqusPlugin.php
<?php
namespace Craft;

class DisqusPlugin extends BasePlugin
{
    public function getName()
    {
        return Craft::t('Disqus');
    }

    public function getDescription()
    {
        return 'Integrates the Disqus commenting system into Craft CMS websites, including Single Sign On (SSO) and custom login/logout URLs.';
    }

    public function getDocumentationUrl()
    {
        return 'https://github.com/nystudio107/disqus/blob/master/README.md';
    }

    public function getReleaseFeedUrl()
    {
        return 'https://raw.githubusercontent.com/nystudio107/disqus/master/releases.json';
    }

    public function getVersion()
    {
        return '1.1.1';
    }

    public function getSchemaVersion()
    {
        return '1.0.0';
    }

    public function getDeveloper()
    {
        return 'nystudio107';
    }

    public function getDeveloperUrl()
    {
        return 'https://nystudio107.com';
    }

    public function hasCpSection()
    {
        return false;
    }

    public function addTwigExtension()
    {
        Craft::import('plugins.disqus.twigextensions.DisqusTwigExtension');

        return new DisqusTwigExtension();
    }

    /**
     * @inheritdoc
     */
    protected function defineSettings()
    {
        return array(
            'disqusShortname' => array(
                AttributeType::String,
                'label' => 'Disqus Site Short Name',
                'default' => ''
            ),
            'useSSO' => array(
                AttributeType::Bool,
                'label' => 'Use Single Sign On',
                'default' => false
            ),
            'disqusPublicKey' => array(
                AttributeType::String,
                'label' => 'Disqus Public Key',
                'default' => ''
            ),
            'disqusSecretKey' => array(
                AttributeType::String,
                'label' => 'Disqus Secret Key',
                'default' => ''
            ),
            'customLogin' => array(
                AttributeType::Bool,
                'label' => 'Use Custom Login/Logout URLs',
                'default' => false
            ),
            'loginName' => array(
                AttributeType::String,
                'label' => 'name',
                'default' => ''
            ),
            'loginButton' => array(
                AttributeType::String,
                'label' => 'button',
                'default' => ''
            ),
            'loginIcon' => array(
                AttributeType::String,
                'label' => 'icon',
                'default' => ''
            ),
            'loginUrl' => array(
                AttributeType::String,
                'label' => 'url',
                'default' => ''
            ),
            'loginLogoutUrl' => array(
                AttributeType::String,
                'label' => 'logout',
                'default' => ''
            ),
            'loginWidth' => array(
                AttributeType::String,
                'label' => 'width',
                'default' => '800'
            ),
            'loginHeight' => array(
                AttributeType::String,
                'label' => 'height',
                'default' => '400'
            ),
        );
    }

    /**
     * @inheritdoc
     */
    public function getSettings()
    {
        $settings = parent::getSettings();
        $base = $this->defineSettings();

        foreach ($base as $key => $row) {
            $override = craft()->config->get($key, 'disqus');

            if (!is_null($override) && !empty($override)) {
                $settings->$key = $override;
            }
        }

        return $settings;
    }

    /**
     * @inheritdoc
     */
    public function getSettingsHtml()
    {
        return craft()->templates->render('disqus/settings', array(
           'settings' => $this->getSettings()
        ));
    }
}


services/DisqusService.php
<?php
namespace Craft;

class DisqusService extends BaseApplicationComponent
{
    /**
     * Output the Disqus Tag
     *
     * @param string $disqusIdentifier
     * @param string $disqusTitle
     * @param string $disqusUrl
     * @param string $disqusCategoryId
     * @param string $disqusLanguage
     *
     * @return string
     */
    public function outputEmbedTag(
        $disqusIdentifier = "",
        $disqusTitle = "",
        $disqusUrl = "",
        $disqusCategoryId = "",
        $disqusLanguage = ""
    ) {
        $settings = craft()->plugins->getPlugin('disqus')->getSettings();
        $disqusShortname = $settings['disqusShortname'];

        $vars = array(
            'disqusShortname' => $disqusShortname,
            'disqusIdentifier' => $disqusIdentifier,
            'disqusTitle' => $disqusTitle,
            'disqusUrl' => $disqusUrl,
            'disqusCategoryId' => $disqusCategoryId,
            'disqusLanguage' => $disqusLanguage,
        );
        $vars = array_merge($vars, $this->getSSOVars());
        $result = $this->renderPluginTemplate('disqusEmbedTag', $vars);

        return $result;
    }

    /**
     * Return the SSO vars
     *
     * @return array
     */
    protected function getSSOVars()
    {
        $settings = craft()->plugins->getPlugin('disqus')->getSettings();
        $vars = array(
            'useSSO' => false,
            'useCustomLogin' => false,
        );
        if ($settings['useSSO']) {
            $data = array();

            // Set the data array
            $currentUser = craft()->userSession->user;
            if ($currentUser) {
                $data['id'] = $currentUser->id;
                if (craft()->config->get('useEmailAsUsername')) {
                    $data['username'] = $currentUser->getFullName();
                } else {
                    $data['username'] = $currentUser->username;
                }
                $data['email'] = $currentUser->email;
                $data['avatar'] = $currentUser->getPhotoUrl();
            }

            // Encode the data array and generate the hMac
            $message = base64_encode(json_encode($data));
            $timestamp = time();
            $hMac = $this->disqusHmacSha1(
                $message
                .' '
                .$timestamp,
                $settings['disqusSecretKey']
            );

            // Set the vars for the template
            $vars = array_merge($vars, array(
                'useSSO' => true,
                'message' => $message,
                'hmac' => $hMac,
                'timestamp' => $timestamp,
                'disqusPublicKey' => $settings['disqusPublicKey'],
                ));

            // Render the SSO custom login template
            if ($settings['customLogin']) {
                $vars = array_merge($vars, array(
                    'useCustomLogin' => true,
                    'loginName' => $settings['loginName'],
                    'loginButton' => $settings['loginButton'],
                    'loginIcon' => $settings['loginIcon'],
                    'loginUrl' => $settings['loginUrl'],
                    'loginLogoutUrl' => $settings['loginLogoutUrl'],
                    'loginWidth' => $settings['loginWidth'],
                    'loginHeight' => $settings['loginHeight'],
                    ));
            }
        }

        return $vars;
    }

    /**
     * Render a plugin template
     *
     * @param $templatePath
     * @param $vars
     *
     * @return string
     */
    protected function renderPluginTemplate($templatePath, $vars)
    {
        // Stash the old template path, and set it to our plugin's templates folder
        $oldPath = method_exists(craft()->templates, 'getTemplatesPath')
            ? craft()->templates->getTemplatesPath()
            : craft()->path->getTemplatesPath();
        $newPath = craft()->path->getPluginsPath() . 'disqus/templates';
        method_exists(craft()->templates, 'setTemplatesPath')
            ? craft()->templates->setTemplatesPath($newPath)
            : craft()->path->setTemplatesPath($newPath);

        // Render the template with our vars passed in
        try {
            $htmlText = craft()->templates->render($templatePath, $vars);
        } catch (\Exception $e) {
            $htmlText = 'Error rendering template ' . $templatePath . ' -> ' . $e->getMessage();
            DisqusPlugin::log($htmlText, LogLevel::Error);
        }

        // Restore the old template path
        method_exists(craft()->templates, 'setTemplatesPath')
            ? craft()->templates->setTemplatesPath($oldPath)
            : craft()->path->setTemplatesPath($oldPath);

        return TemplateHelper::getRaw($htmlText);
    }

    /**
     * HMAC->SHA1
     * From: https://github.com/disqus/DISQUS-API-Recipes/blob/master/sso/php/sso.php
     *
     * @param $data
     * @param $key
     *
     * @return string
     */
    protected function disqusHmacSha1($data, $key)
    {
        $blockSize = 64;
        $hashFunc = 'sha1';
        if (strlen($key) > $blockSize) {
            $key = pack('H*', $hashFunc($key));
        }
        $key = str_pad($key, $blockSize, chr(0x00));
        $iPad = str_repeat(chr(0x36), $blockSize);
        $oPad = str_repeat(chr(0x5c), $blockSize);
        $hMac = pack(
            'H*',
            $hashFunc(
                ($key ^ $oPad).pack(
                    'H*',
                    $hashFunc(
                        ($key ^ $iPad).$data
                    )
                )
            )
        );

        return bin2hex($hMac);
    }
}


twigextensions/DisqusTwigExtension.php
<?php 
namespace Craft;

use Twig_Extension;
use Twig_Filter_Method;
use Twig_Function_Method;

class DisqusTwigExtension extends Twig_Extension
{

    /**
     * @inheritdoc
     */
    public function getName()
    {
        return 'Disqus';
    }

    /**
     * @inheritdoc
     */
    public function getFilters()
    {
        return array(
            'disqusEmbed' => new Twig_Filter_Method($this, 'disqusEmbed'),
        );
    }

    /**
     * @inheritdoc
     */
    public function getFunctions()
    {
        return array(
            'disqusEmbed' => new Twig_Function_Method($this, 'disqusEmbed'),
        );
    }

    /**
     * @param string $disqusIdentifier
     * @param string $disqusTitle
     * @param string $disqusUrl
     * @param string $disqusCategoryId
     * @param string $disqusLanguage
     *
     * @return mixed
     */
    public function disqusEmbed(
        $disqusIdentifier = "",
        $disqusTitle = "",
        $disqusUrl = "",
        $disqusCategoryId = "",
        $disqusLanguage = ""
    ) {
        return craft()->disqus->outputEmbedTag(
            $disqusIdentifier,
            $disqusTitle,
            $disqusUrl,
            $disqusCategoryId,
            $disqusLanguage
        );
    }
}


services/Disqus_UtilsService.php
<?php
namespace Craft;

class Disqus_UtilsService extends BaseApplicationComponent
{

/* --------------------------------------------------------------------------------
    Output the Disqus SSO Tag
-------------------------------------------------------------------------------- */

    public function outputSSOTag($userId = 0)
    {
        $result = "";
        $settings = craft()->plugins->getPlugin('disqus')->getSettings();
        $data = array();

        $currentUser = craft()->userSession->user;
        if ($currentUser)
        {
            $data['id'] = $currentUser->id;
            if (craft()->config->get('useEmailAsUsername'))
                $data['username'] = $currentUser->getFullName();
            else
                $data['username'] = $currentUser->username;
            $data['email'] = $currentUser->email;
            $data['avatar'] = $currentUser->getPhotoUrl();
        }

        $message = base64_encode(json_encode($data));
        $timestamp = time();
        $hmac = dsq_hmacsha1($message . ' ' . $timestamp, $settings['disqusSecretKey']);

        if ($settings['customLogin'])
        {
            $disqusPublicKey = $settings['disqusPublicKey'];
            $loginName = $settings['loginName'];
            $loginButton = $settings['loginButton'];
            $loginIcon = $settings['loginIcon'];
            $loginUrl = $settings['loginUrl'];
            $loginLogoutUrl = $settings['loginLogoutUrl'];
            $loginWidth = $settings['loginWidth'];
            $loginHeight = $settings['loginHeight'];

            echo <<<ENDBLOCK
<script data-cfasync="false" type="text/javascript">
var disqus_config = function() {
    this.page.remote_auth_s3 = "$message $hmac $timestamp";
    this.page.api_key = "$disqusPublicKey";

    this.sso = {
          name:   "$loginName",
          button: "$loginButton",
          icon:   "$loginIcon",
          url:    "$loginUrl",
          logout: "$loginLogoutUrl",
          width:  "$loginWidth",
          height: "$loginHeight"
    };
};
</script>
ENDBLOCK;
        }
        else
        {
            $disqusPublicKey = $settings['disqusPublicKey'];
            echo <<<ENDBLOCK
<script data-cfasync="false" type="text/javascript">
var disqus_config = function() {
    this.page.remote_auth_s3 = "$message $hmac $timestamp";
    this.page.api_key = "$disqusPublicKey";
};
</script>
ENDBLOCK;
        }
        return $result;
    } /* -- outputSSOTag */

/* --------------------------------------------------------------------------------
    Output the Disqus Tag
-------------------------------------------------------------------------------- */

    public function outputEmbedTag($disqusIdentifier = "", $disqusTitle = "", $disqusUrl = "", $disqusCategoryId = "")
    {
        $result = "";
        $settings = craft()->plugins->getPlugin('disqus')->getSettings();
        $disqusTitle = json_encode($disqusTitle);
        if ($settings['useSSO'])
            $this->outputSSOTag();

        $disqusShortname = $settings['disqusShortname'];
        echo <<<ENDBLOCK
<div id="disqus_thread"></div>
<script data-cfasync="false" type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = '$disqusShortname';
    var disqus_identifier = '$disqusIdentifier';
    var disqus_title = '$disqusTitle';
    var disqus_url = '$disqusUrl';
    var disqus_category_id = '$disqusCategoryId';

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';

        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>

<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
ENDBLOCK;

        return $result;
    } /* -- outputEmbedTag */

} /* -- Disqus_UtilsService */

/* --------------------------------------------------------------------------------
    HMAC->SHA1
    From: https://github.com/disqus/DISQUS-API-Recipes/blob/master/sso/php/sso.php
-------------------------------------------------------------------------------- */

    function dsq_hmacsha1($data, $key) {
        $blocksize=64;
        $hashfunc='sha1';
        if (strlen($key)>$blocksize)
            $key=pack('H*', $hashfunc($key));
        $key=str_pad($key,$blocksize,chr(0x00));
        $ipad=str_repeat(chr(0x36),$blocksize);
        $opad=str_repeat(chr(0x5c),$blocksize);
        $hmac = pack(
                    'H*',$hashfunc(
                        ($key^$opad).pack(
                            'H*',$hashfunc(
                                ($key^$ipad).$data
                            )
                        )
                    )
                );
        return bin2hex($hmac);
    } /* -- dsq_hmacsha1 */
